contracts/liquidator/OnChainLiquidator.sol

```diff
@@ -0,0 +1,535 @@
+        address payable recipient;
+        bool toInternalBalance;
+    }
+
+    function batchSwap(
+        SwapKind kind,
+        BatchSwapStep[] memory swaps,
+        IAsset[] memory assets,
+        FundManagement memory funds,
+        int256[] memory limits,
+        uint256 deadline
+    ) external payable returns (int256[] memory);
+}
+
+// Curve Interfaces
+interface ICurveRegistry {
+    function find_pool_for_coins(address _from, address _to) external returns (address);
+}
+
+interface IStableSwap {
+    function exchange(int128 i, int128 j, uint256 _dx, uint256 _min_dy) external payable returns (uint256);
+}
+
+/**
+ * @title Compound's on-chain liquidation contract
+ * @author Compound
+ */
+contract OnChainLiquidator is IUniswapV3FlashCallback, PeripheryImmutableState, PeripheryPayments {
+    /** Errors */
+    error InsufficientAmountOut(address tokenIn, address tokenOut, uint256 amountIn, uint256 amountOut, uint256 amountOutMin, PoolConfig poolConfig);
+    error InvalidArgument();
+    error InsufficientBalance(uint256 available, uint256 required);
+    error InvalidExchange();
+    error InvalidPoolConfig(address swapToken, PoolConfig poolConfig);
+    error Unauthorized();
+
+    /** Events **/
+    event Absorb(address indexed initiator, address[] accounts);
+    event AbsorbWithoutBuyingCollateral();
+    event BuyAndSwap(address indexed tokenIn, address indexed tokenOut, uint256 baseAmountPaid, uint256 assetBalance, uint256 amountOut);
+    event Pay(address indexed token, address indexed payer, address indexed recipient, uint256 value);
+    event Swap(address indexed tokenIn, address indexed tokenOut, uint256 amountIn, uint256 amountOut, PoolConfig poolConfig);
+
+    enum Exchange {
+        Uniswap,
+        SushiSwap,
+        Balancer,
+        Curve
+    }
+
+    // XXX make this less gassy; rearrange fields
+    struct PoolConfig {
+        Exchange exchange;
+        uint24 uniswapPoolFee;
+        bool swapViaWeth;
+        bytes32 balancerPoolId;
+    }
+
+    /** Liquidator configuration constants **/
+
+    /// @notice Uniswap router used for token exchange
+    address public constant uniswapRouter = 0xE592427A0AEce92De3Edee1F18E0157C05861564;
+
+    /// @notice SushiSwap router used for token exchange
+    address public constant sushiSwapRouter = 0xd9e1cE17f2641f24aE83637ab66a2cca9C378B9F;
+
+    /// @notice The scale for asset price calculations
+    uint256 public constant QUOTE_PRICE_SCALE = 1e18;
+
+    /// @notice The asset pool configurations
+    mapping(address => PoolConfig) public poolConfigs;
+
+    /**
+     * @notice Construct a new liquidator instance
+     * @param _factory Address of Uniswap V3 Factory
+     * @param _WETH9 Address of WETH9
+     **/
+    constructor(address _factory, address _WETH9) PeripheryImmutableState(_factory, _WETH9) {}
+
+    /**
+     * @dev Returns lesser of two values
+     */
+    function min(uint256 a, uint256 b) internal view returns (uint256) {
```

Comment:
nit: recommended by OZ to put `internal` fns after `external` ones

Reply:
reordered